### PR TITLE
ACKS2: fix for ill-conditioned Jacobi preconditioner cases

### PR DESCRIPTION
**Pull request by ohearnk**
TODO: code included for non-Kokkos ACKS2 code but needs to be added for Kokkos code.

**Summary**

This MR contains code changes to check the conditioning of the Jacobi preconditioner used in the BiCGStab solver for charges in ACKS2.   If ill-conditioning is detected, preconditining is disabled for the ACKS2 region to make the preconditioner well-conditioned.

**Related Issue(s)**

Solver convergence issues have been observed with some recent O-H test cases.  CC: @stanmoore1

**Author(s)**

Kurt A. O'Hearn, Michigan State University <ohearnku@msu.edu>

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

N/A

**Implementation Notes**

For Jacobi (i.e., diagonal inverse) preconditioning in the iterative linear solvers (i.e., BiCGStab), the singular values are the diagonal entries themselves (in absolute value).  So, directly computing the condition number of the preconditioner is straightforward.

**Post Submission Checklist**

- [ ] The feature or features in this pull request is complete
- [ X] Licensing information is complete
- [ X] Corresponding author information is complete
- [ X] The source code follows the LAMMPS formatting guidelines
- [ X] The feature has been verified to work with the conventional build system
- [ X] The feature has been verified to work with the CMake based build system

### CODE REVIEW

**Comment by stanmoore1** [comment]
@ohearnk thanks I will port this to Kokkos

File: src/REAXFF/fix_acks2_reaxff.cpp [hunk 1]
```diff
@@ -396,6 +396,34 @@ void FixACKS2ReaxFF::init_matvec()
     }
   }
 
+  /* compute condition number of Jacobi preconditioner */
+  double sing_max = std::numeric_limits<double>::min();
```
Comment (akohlmey) [inline src/REAXFF/fix_acks2_reaxff.cpp:402]: To use these function templates, `#include <limits>` should be added in the beginning of the file.

**Comment by akohlmey** [comment]
@ohearnk when I run this branch on the examples, the QeQ calculation is significantly slowed down, presumably due to more iterations, and for the `in.water.acks2.field` input example it is particularly extreme with multiple error messages that convergence was not achieved after 1000 iterations. The percentage of time spent on QeQ increases from almost 60% in the reference to over 90% with this branch.

**Comment by stanmoore1** [comment]
> @ohearnk when I run this branch on the examples, the QeQ calculation is significantly slowed down, presumably due to more iterations, and for the `in.water.acks2.field` input example it is particularly extreme with multiple error messages that convergence was not achieved after 1000 iterations. The percentage of time spent on QeQ increases from almost 60% in the reference to over 90% with this branch.

@ohearnk we may need to set the threshold for the condition number much higher than `100.0`, i.e. just below where the problematic O-H test case still passes, in order to recover performance for non-problematic cases.

**Comment by akohlmey** [comment]
@ohearnk @stanmoore1 this has been sitting inactive for multiple months now. I will close the pull request now. Please feel free to submit a new one, if you have completed the work.

**Pull request closed without merging**